design/keycloak.x/storage-persistence.md

```diff
@@ -0,0 +1,532 @@
+changes the logical object version. In contrast, adding a method to model
+that operates on existing fields does not change model version.
+
+For example, [this change in `RoleModel`](https://github.com/keycloak/keycloak/commit/9ef4c7fffd75ed63c49713405083c17e643ff80a#diff-38a40e77fbc7b5c42eb6cda54cf5cd6e)
+added attributes to the roles and thus means update to the logical object version.
+On the other hand, [this change in `RoleModel`](https://github.com/keycloak/keycloak/commit/8f7f9e0b9c1ea0a50a4f6292ffa09c856e067c10#diff-38a40e77fbc7b5c42eb6cda54cf5cd6e)
+did operates on data already available in the model, thus the logical object version
+of `RoleModel` remains unchanged.
+
+For each of the logical entities (realm, user, client etc.),
+the _current store version_ is the logical object version that
+the store uses for writing objects.
+
+Conditions for the readability of an object stored by store of version _N_
+are as follows:
+
+1. _Backward compatibility._
+   Object stored by store of version _N_ MUST be readable by any newer store,
+   i.e. store of version _M_ where _M ≥ N_
+
+2. _Forward compatibility._
+   Object stored by store of version _N_ MUST be readable by a previous store,
+   i.e. store of version _N-1_.
+   This is to support seamless upgrade in a manner where nodes
+   of version _N-1_ in the cluster are gradually replaced
+   by newer version. It also provides means to revert to
+   the previous version if anything goes wrong.
+
+3. _Extended forward compatibility_ (optional).
+   Object stored by store of version _N_ MAY be readable by much older store,
+   i.e. store of version _M_ where _M ≤ N-2_, but the store MUST
+   throw an `IllegalArgumentException` if the object cannot
+   be reliably reconstructed from the stored data.
+
+The following matrix illustrates the above conditions. It shows whether
+an object stored in version _X_ in a store of version _Y_ is readable
+(green) or unreadable (red):
+
+![Matrix shows whether a stored object in format _X_ is readable (green) or unreadable (red) by store of version _Y_](img/storage-persistence-read-compatibility.svg)
+
+For the sake of simplicity, object versions are treated as a sequence of integers
+thorough this text. See [below](#_otherVersioning) for other options.
+
+Versions of individual stores in the same server version are unrelated.
+For example, in Keycloak of version 1.2.3, versions can be 53 for user objects,
+41 for clients, and 14 for realms. Versioning of Keycloak and the logical objects
+is independent, so it is possible that the Keycloak version 1.2.4 may have
+the same object versions as in 1.2.3, that one of them has changed,
+or all have changed.
+
+*Note on testability:* It must be possible to automatically test that
+a store implementation satisfies all the requirements stated in this section.
+For this purpose, there is
+[a project in keycloak-playground](https://github.com/keycloak/keycloak-playground/no-downtime-upgrade/) repository
```

Comment:
@hmlnarik Please, replace current playground url with the valid one: https://github.com/keycloak/keycloak-playground/tree/master/no-downtime-upgrade